[FLINK-13277][hive] add documentation of Hive source/sink

lirui-apache opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

To add documentation for hive connector.


## Brief change log

  - Added doc for hive connector.

## Verifying this change

Doesn't need testing.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

lirui-apache commented:
Hive table source and sink are not mentioned in the doc since they're not meant to be public APIs.
@xuefuz @zjuwangg please let me know if you find anything to add.

xuefuz commented:
@lirui-apache Could you restart CI test? Previous failure doesn't seem related.

bowenli86 commented on docs/dev/batch/connectors.md:
```diff
@@ -217,4 +217,21 @@ The example shows how to access an Azure table and turn data into Flink's `DataS
+2. Inserting into partitioned tables is not supported.
+3. ACID tables are not supported.
+4. Bucketed tables are not supported.
+5. UNION type is not supported.
```
we can move the type related limitations to https://github.com/apache/flink/blob/master/docs/dev/table/hive/index.md

bowenli86 commented on docs/dev/batch/connectors.md:
```diff
@@ -217,4 +217,21 @@ The example shows how to access an Azure table and turn data into Flink's `DataS
 
 This [GitHub repository documents how to use MongoDB with Apache Flink (starting from 0.7-incubating)](https://github.com/okkam-it/flink-mongodb-test).
 
+## Hive Connector
```
maybe there should be an additional cross page link with https://github.com/apache/flink/blob/master/docs/dev/table/hive/read_write_hive.md except that with catalog.md?

bowenli86 commented on docs/dev/batch/connectors.md:
```diff
@@ -217,4 +217,21 @@ The example shows how to access an Azure table and turn data into Flink's `DataS
+After HiveCatalog is setup, you can use Table API or Flink SQL Client to access tables in Apache Hive.
+Same as HiveCatalog, the officially supported Apache Hive versions are 2.3.4 and 1.2.1.
+
+### Limitations
```
need to consolidate with https://github.com/apache/flink/blob/master/docs/dev/table/hive/read_write_hive.md#limitations

lirui-apache commented:
Given that we already have `read_write_hive.md`, maybe we don't need a section for Hive connector in `connectors.md`. What do you guys think?

xuefuz commented:
I think we can consolidate the content in the PR to the relevant content in read_write_hive.md and put a cross reference in connectors.md.

lirui-apache commented:
@bowenli86 @xuefuz I have updated the PR to address your comments. And since we have decided not to support timestamp in 1.9.0, I also added that to the limitations part. Please have a look. Thanks.

xuefuz commented on docs/dev/table/hive/read_write_hive.md:
```diff
@@ -120,4 +120,11 @@ Flink SQL> INSERT INTO mytable (name, value) VALUES ('Tom', 4.72);
+3. ACID tables are not supported.
+4. Bucketed tables are not supported.
+5. Some data types are not supported. See the [limitations]({{ site.baseurl }}/dev/table/hive/#limitations) for details.
+6. Only a limited number of table storage formats have been tested, namely text, ORC, and Parquet.
```
Is squencefile tested?

xuefuz commented on docs/dev/batch/connectors.md:
```diff
@@ -217,4 +217,10 @@ The example shows how to access an Azure table and turn data into Flink's `DataS
+## Hive Connector
+
+Starting from 1.9.0, Apache Flink provides Hive connector to access Apache Hive tables. [HiveCatalog]({{ site.baseurl }}/dev/table/catalogs.html#hivecatalog) is required in order to use the Hive connector.
+After HiveCatalog is setup, please refer to [Reading & Writing Hive Tables]({{ site.baseurl }}/dev/table/hive/read_write_hive.html) for how to use the Hive connector and its limitations.
```
Nit: Reword it like "for the usage of the Hive connector and its limitations" to be more clear.

bowenli86 commented on docs/dev/table/hive/index.md:
```diff
@@ -263,3 +259,5 @@ The following limitations in Hive's data types impact the mapping between Flink
 * Flink's `TIMESTAMP_WITH_TIME_ZONE` is not supported by Hive
 * Flink's `TIMESTAMP_WITH_LOCAL_TIME_ZONE` is not supported by Hive
 * Flink's `MULTISET` is not supported by Hive
+* Hive's `UNION` type is not supported
```
I'll make these couple changes of data types in a different PR so this PR can focus on Hive source/sink. Thanks!

bowenli86 commented on docs/dev/table/hive/read_write_hive.md:
```diff
@@ -120,4 +120,11 @@ Flink SQL> INSERT INTO mytable (name, value) VALUES ('Tom', 4.72);
 
 ### Limitations
```
one more limitations: cannot read Hive views due to limitations of planner

bowenli86 commented on docs/dev/table/hive/read_write_hive.md:
```diff
@@ -120,4 +120,11 @@ Flink SQL> INSERT INTO mytable (name, value) VALUES ('Tom', 4.72);
+3. ACID tables are not supported.
+4. Bucketed tables are not supported.
+5. Some data types are not supported. See the [limitations]({{ site.baseurl }}/dev/table/hive/#limitations) for details.
+6. Only a limited number of table storage formats have been tested, namely text, ORC, and Parquet.
```
do we need extra dependencies if users use ORC/Parquet?

A user trying to use lzo found that he needed to include hadoop-lzo himself
http://apache-flink.147419.n8.nabble.com/flink1-10-hive-td336.html

lirui-apache replied:
I don't think so according to our previous tests. Also manually verified reading ORC/Parquet tables with different compression codec and seems we're good.

lirui-apache commented:
@xuefuz @bowenli86 PR updated according to your feedbacks.

bowenli86 commented:
btw, we need to copy the exact same content from xxx.md to xxx.zh.md, to keep English and Chinese docs in sync and facilitate translation in the future

lirui-apache commented:
@bowenli86 Thanks for letting me know. I have synced the changes to zh docs.

bowenli86 commented:
LGTM, merging
